sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2466,20 +2489,26 @@ class Dataset[T] private[sql](
 
     field.dataType match {
       case IntegerType | LongType | DoubleType | FloatType | BooleanType | ByteType =>
-        val validity = allocator.buffer(numBytesOfBitmap(numOfRows))
+        val validityVector = new BitVector("validity", allocator)
+        val validityMutator = validityVector.getMutator()
+        validityVector.allocateNew(numOfRows)
+        validityMutator.setValueCount(numOfRows)
         val buf = allocator.buffer(numOfRows * field.dataType.defaultSize)
         var nullCount = 0
-        rows.foreach { row =>
+        rows.zipWithIndex.foreach { case (row, index) =>
```

Comment:
I think it would be better just use a while loop here since `zipWithIndex` will iterate and copy the items in an array
```
var index = 0
while (index < rows.length) {
  ..
  index += 1
}